modules/packages/LinearAlgebra.chpl

```diff
@@ -2064,6 +2319,123 @@ proc kron(A: [?ADom] ?eltType, B: [?BDom] eltType) {
 }
 
 
+/*
+  Matrix exponential using Pade approximation
+
+  Params:
+  A: Expects an N*N square matrix
+  use_exact_onenorm: boolean value specifying if the onenorm has to be exact.
+  Defaults to true.
+
+  TODO: Need to write a method to find approximate norm.
+  Currently only supports for use_exact_onenorm = true.
+
+  Returns:
+  expA: N*N matrix which is Matrix exponential of `A`
+
+  References:
+  [1] Awad H. Al-Mohy and Nicholas J. Higham (2009)
+  "A New Scaling and Squaring Algorithm for the Matrix Exponential."
+  SIAM Journal on Matrix Analysis and Applications.
+  31 (3). pp. 970-989. ISSN 1095-7162
+*/
+
+proc expm(A: [], param use_exact_onenorm=true){
+
+  if A.rank != 2 || A.domain.dim(0) != A.domain.dim(1) then {
+    writeln('Square Matrix Expected');
+    return A;
+  }
+
+  var h = new ExpmPadeHelper(A, use_exact_onenorm);
+
+  // Try Pade order 3
+  var eta_1 = max(h.comp_Loosed4(), h.comp_Loosed6());
+  if eta_1 < 1.495585217958292e-002 then {
+    var mat = h.pade3();
+    var U = mat[0];
+    var V = mat[1];
+    return solvePQ(U, V);
+  }
+
+  // Try Pade order 5
+  var eta_2 = max(h.comp_Exactd4(), h.comp_Loosed6());
+  if eta_2 < 2.539398330063230e-001 then {
+    var mat = h.pade5();
+    var U = mat[0];
+    var V = mat[1];
+    return solvePQ(U, V);
+  }
+
+  // Try Pade order 7 and 9
+  var eta_3 = max(h.comp_Exactd6(), h.comp_Loosed8());
+  if eta_3 < 9.504178996162932e-001 then {
+    var mat = h.pade7();
+    var U = mat[0];
+    var V = mat[1];
+    return solvePQ(U, V);
+  }
+
+  if eta_3 < 2.097847961257068e+000 then {
+    var mat = h.pade9();
+    var U = mat[0];
+    var V = mat[1];
+    return solvePQ(U, V);
+  }
+
+  // Use Pade order 13.
+  var eta_4 = max(h.comp_Loosed8(), h.comp_Loosed10());
+  var eta_5 = min(eta_3, eta_4);
+  var theta_13 = 4.25;
+
+  var s = 0;
+  // choose smallest s such that s>=0 and 2**(-s) eta_5 <= theta_13
+  if eta_5 == 0 then {
+    // Nilpotent special case
+    s = 0;
+  }
+  else {
+    s = max(ceil(log2(eta_5 / theta_13)): int, 0);
+  }
+  var mat = h.pade13_scaled(s);
+  var U = mat[0];
+  var V = mat[1];
+  var X = solvePQ(U, V);
+  // According to the paper X = r_13(A)^(2^s): achieved by repeated squaring.
+  for i in 1..s {
+    X = dot(X, X);
+  }
+  return X;
+}
+
+/*
+This method finds P Q Matrices, where
+P = U + V
+Q = -U + V
+it then returns X where Q*X = P
+*/
+private proc solvePQ(U: [], V: []){
+  var D = U.indices;
+  var P = reshape(U + V, D);
+  var Q = reshape(-U + V, D);
+  var res : [D] real;
+  var b : [D.dim(0)] real;
+
+  // Need to rewrite solve function
```

Comment:
I'm tempted to directly call the LAPACK routine here, since that has all of this built in. We should capture this as an issue though (I thought one existed, but I could not find it).

Reply:
We should remember to file a bug report here.

Reply:
Filed a bug [17912](https://github.com/chapel-lang/chapel/issues/17912)